src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -170,9 +184,50 @@ public JSONObject getMatchingCourses(HttpServletRequest request) {
       }
     }
     List<Entity> results = db.prepare(courseQuery).asList(FetchOptions.Builder.withDefaults());
+
     return results;
   }
 
+  /* Does substring matching through regex. */
+  private List<Entity> processResults(HttpServletRequest request, List<Entity> fuzzyResults) {
+    List<Entity> processed = new ArrayList<>();
+    Boolean hasName = !request.getParameter("course-name").isEmpty();
+    Boolean hasProf = !request.getParameter("prof-name").isEmpty();
+    Pattern namePattern = Pattern.compile("/(?=a)b/"); // Always false.
+    Pattern profPattern = Pattern.compile("/(?=a)b/"); // Always false.
+    if (hasName) {
+      String name = request.getParameter("course-name").toLowerCase();
+      namePattern = Pattern.compile("^.*" + name + ".*$");
+    }
+
+    if (hasProf) {
+      String prof = request.getParameter("prof-name").toLowerCase();
+      profPattern = Pattern.compile("^.*" + prof + ".*$");
+    }
+    Boolean nameMatches;
+    Boolean profMatches;
+    for (Entity e : fuzzyResults) {
+      String searchName = (String) e.getProperty("search-name");
+      String searchProf = (String) e.getProperty("search-professor");
+      if (hasName) {
+        Matcher nameMatcher = namePattern.matcher(searchName);
+        nameMatches = nameMatcher.matches();
+      } else {
+        nameMatches = true;
+      }
+      if (hasProf) {
+        Matcher profMatcher = profPattern.matcher(searchProf);
+        profMatches = profMatcher.matches();
+      } else {
+        profMatches = true;
+      }
+      if (nameMatches && profMatches && (hasName || hasProf)) {
```

Comment:
Do you need to check if the request had a name or prof?

Reply:
if it didn't have either, then nameMatches && profMatches would be true, but I don't want to add it to the list. So I want to make sure that both of them aren't true just because the parameter didn't exist.

Reply:
Why don't you want to add to the list if the search query had no name or prof?